design-documents/database/declarative-schema-database-compatibility.md

```diff
@@ -0,0 +1,207 @@
+# Declarative schema database compatibility
+
+## About
+
+Declarative schema relies on querying metadata from the DB.
+The query itself and format of the result may differ for different RDBMS systems.
+For example, here is the query from `Magento\Framework\Setup\Declaration\Schema\Db\MySQL\DbSchemaReader::readColumns`
+
+```
+SELECT 
+    COLUMN_NAME as 'name',
+    COLUMN_DEFAULT as 'default',
+    DATA_TYPE as 'type',
+    IS_NULLABLE as 'nullable',
+    COLUMN_TYPE as 'definition',
+    EXTRA as 'extra',
+    COLUMN_COMMENT as 'comment'
+FROM
+    information_schema.COLUMNS
+WHERE
+    TABLE_SCHEMA = 'magento'
+AND
+    TABLE_NAME = 'store_website'
+ORDER BY
+    ORDINAL_POSITION ASC;
+```
+
+For `COLUMN_DEFAULT`, PHP PDO connector returns `NULL` for MySQL and `"NULL"` (string) for MariaDB 10.2.7+.
+
+## Design
+
+Introduce new interfaces under Db namespace that would allow to get information about table schema, refactor implementations of `\Magento\Framework\Setup\Declaration\Schema\Db\DbSchemaReaderInterface` in declarative schema to use interfaces under Db namespace.
+
+Revise implementation of these interfaces to see if anything need to be changed here
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DDLTriggerInterface`
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DbDefinitionProcessorInterface` (see usages of `\Magento\Framework\Setup\Declaration\Schema\Db\DefinitionAggregator`)
+
+New interfaces and types for managing table schema.
+
+```
+class InformationSchema\Table
+{
+    public function getSchema(): string {}
+    public function getName(): string {}
+    public function getEngine(): string {}
+    public function getComment(): string {}
+    public function getCollation(): string {}
+    public function getCharset(): string {}
+}
+```
+Note: auto increment is omitted intentionally. Auto increment should not be used for the application logic.
+
+```
+class InformationSchema\Extra
+{
+    public function getOnUpdate(): string {}
+}
+```
+
+```
+abstract class InformationSchema\Table\Column
+{
+    public function getTableName(): string {}
+    public function getName(): string {}
+    public function getIsNullable(): bool {}
+    public function getExtra(): Extra {}
+    public function getComment() {}
+    public function getDefaultValue(): ?string {}
+}
+```
+
+```
+class InformationSchema\Table\VarcharColumn extends Column
+{
+    public const TYPE = 'VARCHAR';
+
+    public function getCharLength(): int {}
+    public function getCharsetName(): string {}
+    public function getCollationName(): string {}
+}
+```
+
+```
+class InformationSchema\Table\IntColumn extends Column
+{
+    public const TYPE = 'INT';
+
+    public function isUnsigned(): bool {}
+    public function getPadding(): int {}
+    public function getType(): IntEnum {}
```

Comment:
Propose to have different types instead.

Reply:
Agreed